### PR TITLE
dhcpv6 packet handling code refine

### PR DESCRIPTION
**Pull request by jcaiMR**
#Description:

Fix https://msazure.visualstudio.com/One/_workitems/edit/16590224/

1. Define DHCPv6 option, DHCPv6 message, DHCPv6 Relay message marshal/un-marshal class
2. Remove original static send buffer, send buffer will from message class (memory from heap)
3. code refine

#how you test it
dhcpv6 related scripts running.

### CODE REVIEW

File: src/relay.cpp [hunk 1]
```diff
@@ -71,6 +71,153 @@ std::map<int, std::string> counterMap = {{DHCPv6_MESSAGE_TYPE_UNKNOWN, "Unknown"
...
+
+bool Options::Add(OptionCode key, const uint8_t *value, uint16_t len) {
+    std::vector<uint8_t> option_v(value, value + len);
+    options[key] = option_v;
+    return true;
+}
+
+bool Options::Delete(OptionCode key) {
+    if (options.find(key) == options.end()) {
+        return false;
```
Comment (yxieca) [inline src/relay.cpp:85]: What is the reason for returning false here?
What is the harm of deleting non-existing key?

File: src/relay.cpp [hunk 2]
```diff
@@ -71,6 +71,153 @@ std::map<int, std::string> counterMap = {{DHCPv6_MESSAGE_TYPE_UNKNOWN, "Unknown"
...
+        return std::vector<uint8_t>{};
+    }
+    return options[key];
+}
+
+std::vector<uint8_t> *Options::MarshalBinary() {
+    if (options.empty()) {
+        return nullptr;
+    }
+    list.clear();
```
Comment (yxieca) [inline src/relay.cpp:102]: Is there a risk of using 'list' recursively?

File: src/relay.cpp [hunk 3]
```diff
@@ -71,6 +71,153 @@ std::map<int, std::string> counterMap = {{DHCPv6_MESSAGE_TYPE_UNKNOWN, "Unknown"
...
+        list.push_back(((uint8_t *)&type)[1]);
+        list.push_back(((uint8_t *)&length)[0]);
+        list.push_back(((uint8_t *)&length)[1]);
+        list.insert(list.end(), itr.second.begin(), itr.second.end());
+    }
+    return &list;
+}
+
+bool Options::UnmarshalBinary(const uint8_t *packet, uint16_t length) {
+    while (length >= sizeof(dhcpv6_option)) {
```
Comment (yxieca) [inline src/relay.cpp:116]: Should we log/warn if there is residue bytes that are less than an option size?

File: src/relay.cpp [hunk 4]
```diff
@@ -71,6 +71,153 @@ std::map<int, std::string> counterMap = {{DHCPv6_MESSAGE_TYPE_UNKNOWN, "Unknown"
...
+    if (buffer == nullptr) {
+        buffer.reset(new (std::nothrow)uint8_t[BUFFER_SIZE]);
+        if (!buffer) {
+            syslog(LOG_ERR, "Failed to init relay msg buffer\n");
+            exit(1);
+        }
+    }
+
+    auto ptr = buffer.get();
+    std::memcpy(ptr, &hdr, sizeof(dhcpv6_relay_msg));
```
Comment (yxieca) [inline src/relay.cpp:150]: hdr as a member attribute is not explicit. Can you make it more explicit that this is a member attribute? This comments is applicable to other member attributes in this PR too.

File: src/relay.cpp (hunk 1 above)
Comment (jcaiMR) [reply src/relay.cpp:85]: Delete non-existing key is no harm, I make it return false in order for a notification to callers in future. 
This place return false will not impart option unmarshal/marshal operation.

File: src/relay.cpp (hunk 2 above)
Comment (jcaiMR) [reply src/relay.cpp:102]: Yes, good point. I will add some protection for this case. And update Class member function descriptions.

File: src/relay.cpp (hunk 3 above)
Comment (jcaiMR) [reply src/relay.cpp:116]: Sure, I will update in next diff. Thanks for suggestion.

File: src/relay.cpp (hunk 4 above)
Comment (jcaiMR) [reply src/relay.cpp:150]: Got it, will change member name to use '_' as prefix.
Thanks Ying for comments.

Comment (yxieca) [reply src/relay.cpp:150]: What is the common coding style to mark a member attribute in SONIC repos? '_" or 'm_'? Can you check other C++ code repos?

File: src/relay.cpp [hunk 5]
```diff
@@ -71,6 +71,153 @@ std::map<int, std::string> counterMap = {{DHCPv6_MESSAGE_TYPE_UNKNOWN, "Unknown"
...
+bool RelayMsg::UnmarshalBinary(const uint8_t *packet, uint16_t len) {
+    if (len < sizeof(dhcpv6_relay_msg)) {
+        syslog(LOG_WARNING, "unmarshal relay msg error, invalid packet length %d", len);
+        return false;
+    }
+    auto msg_hdr = (dhcpv6_relay_msg *)packet;
+    hdr.msg_type = msg_hdr->msg_type;
+    hdr.hop_count = msg_hdr->hop_count;
+    std::memcpy(&hdr.link_address, &msg_hdr->link_address, sizeof(struct in6_addr));
+    std::memcpy(&hdr.peer_address, &msg_hdr->peer_address, sizeof(struct in6_addr));
```
Comment (yxieca) [inline src/relay.cpp:171]: What is the benefit of using std::memcpy v.s. variable assignment?

File: src/relay.cpp (hunk 4 above)
Comment (jcaiMR) [reply src/relay.cpp:150]: Got it, checked in other SONIC repo. Private member use "m_" as prefix. Public member/methods using Camel-Case naming. Will apply in latest diff.

File: src/relay.cpp [hunk 6]
```diff
@@ -517,30 +665,37 @@ void relay_client(int sock, const uint8_t *msg, int32_t len, const ip6_hdr *ip_h
...
-    dhcpv6_relay_msg new_message;
-    auto current_buffer_position = buffer;
     auto dhcp_relay_header = parse_dhcpv6_relay(msg);
 
-    if (dhcp_relay_header->hop_count >= HOP_LIMIT)
+    if (dhcp_relay_header->hop_count >= HOP_LIMIT) {
+        char addr_str[INET6_ADDRSTRLEN];
+        inet_ntop(AF_INET6, &ip_hdr->ip6_src, addr_str, INET6_ADDRSTRLEN);
+        syslog(LOG_INFO, "receive relay-forward message from %s with hop count %d over limit",
+               addr_str, dhcp_relay_header->hop_count);
```
Comment (yxieca) [inline src/relay.cpp:674]: Should the wording be: "dropping relay-forward message"?

File: src/relay.cpp [hunk 7]
```diff
@@ -451,54 +597,56 @@ void prepare_socket(int &local_sock, int &server_sock, relay_config &config) {
...
+        return;
+    }
+    increase_counter(config->state_db, counterVlan.append(config->interface), dhcpv6.msgHdr.msg_type);
 
-        if ((unsigned)(current_buffer_position + sizeof(linklayer_addr_option) - buffer) > sizeof(buffer)) {
-            return;
-        }
-        memcpy(current_buffer_position, &option79, sizeof(linklayer_addr_option));
-        current_buffer_position += sizeof(linklayer_addr_option);
+    /* generage relay packet */
```
Comment (kellyyeh) [inline src/relay.cpp:615]: Typo here

**Comment by kellyyeh** [comment]
LGTM aside from minor comment! Thanks Jie

File: src/relay.cpp (hunk 6 above)
Comment (jcaiMR) [reply src/relay.cpp:674]: sure, will update in latest diff.

File: src/relay.cpp (hunk 7 above)
Comment (jcaiMR) [reply src/relay.cpp:615]: Good catch, will update in latest diff.

**Review by kellyyeh — APPROVED** [review]

**Comment by jcaiMR** [comment]
> Hi @jcaiMR, I couldn't access the link in the description, can you provide more details on the symptoms of the problem?

Hi Vivek, this PR trying to improve DHCP Relay packet handling code.

We want to introduce packet marshal/un-marshal to give a full protection for malformed dhcpv6 packets from client.
DHCPv6 packets from client may has malformed options (wrong TLV difinition), we want to protect this kind of packets not leak this kind of packets to server.

Also we want to have a simple way for options add/delete/modify, which may easy for new options support in future.

**Comment by vivekrnv** [comment]
> > Hi @jcaiMR, I couldn't access the link in the description, can you provide more details on the symptoms of the problem?
> 
> Hi Vivek, this PR trying to improve DHCP Relay packet handling code.
> 
> We want to introduce packet marshal/un-marshal to give a full protection for malformed dhcpv6 packets from client. DHCPv6 packets from client may have malformed options (wrong TLV difinition), we want to protect this kind of packets and not leak this kind of packets to server.
> 
> Also we want to have a simple way for options add/delete/modify, which may easy for new options support in future.

Okay, thanks for clarification. I've never understood why we had to explicitly deal with incoming packet byte by byte. I think this would be much simpler if we use a library like this https://pcapplusplus.github.io/. https://github.com/seladb/PcapPlusPlus/blob/master/Packet++/header/DhcpV6Layer.h

**Comment by jcaiMR** [comment]
> > > Hi @jcaiMR, I couldn't access the link in the description, can you provide more details on the symptoms of the problem?
> > 
> > 
> > Hi Vivek, this PR trying to improve DHCP Relay packet handling code.
> > We want to introduce packet marshal/un-marshal to give a full protection for malformed dhcpv6 packets from client. DHCPv6 packets from client may have malformed options (wrong TLV difinition), we want to protect this kind of packets and not leak this kind of packets to server.
> > Also we want to have a simple way for options add/delete/modify, which may easy for new options support in future.
> 
> Okay, thanks for clarification. I've never understood why we had to explicitly deal with incoming packet byte by byte. I think this would be much simpler if we use a library like this https://pcapplusplus.github.io/. https://github.com/seladb/PcapPlusPlus/blob/master/Packet++/header/DhcpV6Layer.h

PcapPlusPlus also deal with packet byte by byte in TLVRecord Class (you have to go through the TLVs for each packet to find/insert options, right ? ).
PcapPlusPlus stops the walk when find the target option user interested, here we want to walk all options for a full packet check (detect all malformed TLVs).  What we did more is alloc buffer for each options during walk, this may benefit if you want to modify several options at the same time, you don't need walk it again (PcapPlusPlus need). I think this is different consideration from different point of view. 
If you have performance consideration, I think we can improve in future like optimize memory allocation, batch update counter to have less Redis api calls ... 
Thanks Vivek for comments, let's keep touch to make the code better.

**Comment by vivekrnv** [comment]
Thanks for taking time to analyze. I agree that the pcaplus will still have to read the packet byte by byte. My point in using an external library was to let the heavy lifting of the packet processing be handled by the lib which makes the code less error-prone and easy to maintain and read.

Now I'm not super familiar with PcapPlus but having a brief look at the code it seems, the DhcpV6Layer class provides multiple API's to walk-through, add, delete options. And if an option is malformed the `DhcpV6Option::getType` should return DHCPV6_OPT_UNKNOWN as i understand. Wouldn't that let us know if the option is malformed ?https://github.com/seladb/PcapPlusPlus/blob/master/Packet++/src/DhcpV6Layer.cpp#L12

Regarding your comment

> > PcapPlusPlus stops the walk when find the target option user interested, here we want to walk all options for a full packet check (detect all malformed TLVs). What we did more is alloc buffer for each options during walk, this may benefit if you want to modify several options at the same time, you don't need walk it again (PcapPlusPlus need).

If we aren't editing options, i think pcaplus just does one walk over the entire data if we are looping over it option by option, 
https://github.com/seladb/PcapPlusPlus/blob/142e3de80660343f61c8596d0c50f728c9d9c554/Packet%2B%2B/src/DhcpV6Layer.cpp#L154 && https://github.com/seladb/PcapPlusPlus/blob/142e3de80660343f61c8596d0c50f728c9d9c554/Packet%2B%2B/src/DhcpV6Layer.cpp#L159

Modifying multiple options is one use-case where the current solution seem to have an edge, In Pcapplus, i think it does that by modifying options one-by-one. 

Regardless, i think using a standardized library greatly reduces chances of error and make the code much cleaner. So i still think using the external lib might be a better/safe choice, let me know what you think.

**Comment by qiluo-msft** [comment]
@jcaiMR If this is not a bug fix, let's remove the label to cherry-pick 202012.

**Comment by jcaiMR** [comment]
> @jcaiMR If this is not a bug fix, let's remove the label to cherry-pick 202012.

Sure, removed the label.

**Pull request merged**